pkg/watchdog/impl.go

```diff
@@ -0,0 +1,109 @@
+package watchdog
+
+import (
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"net"
+	"time"
+
+	"github.com/Cloud-Foundations/golib/pkg/log"
+)
+
+func (config *Config) setDefaults() {
+	config.ArmTime = time.Second * 10
+	config.CheckInterval = time.Second * 5
+	config.ExitTime = time.Second * 15
+}
+
+func newWatchdog(config Config, logger log.DebugLogger) (*Watchdog, error) {
+	if config.TcpPort < 1 {
+		return nil, errors.New("no TCP port number specified")
+	}
+	if config.ArmTime < time.Second {
+		config.ArmTime = time.Second // Minimum.
+	}
+	if config.CheckInterval < time.Millisecond*100 {
+		config.CheckInterval = time.Millisecond * 100 // Minimum.
+	}
+	if config.ExitTime < config.CheckInterval*2 {
+		config.ExitTime = config.CheckInterval * 2 // Minimum.
+	}
+	w := &Watchdog{
+		addr:   fmt.Sprintf(":%d", config.TcpPort),
+		c:      config,
+		logger: logger,
+	}
+	go w.enter()
+	return w, nil
+}
+
+func (w *Watchdog) check() error {
```

Comment:
I dont like the exit params, it make more sense to be bool, err. Where error is only set when there is an internal error. Which implies the obvious: what do do on internal errors? (for example SetDeadline fails).

Reply:
The error is nice because it allows the caller to log the message rather than having to add logging calls for each error in the `check()` function.
As for what happens on internal error, I think it's better to classify those as check failures so that the programme is more likely to exit. If there are internal errors like this, the whole programme could be experiencing problems.